### ISSUE TITLE
Completed pods are sometimes deleted across node upgrade

### ISSUE THREAD

**Issue by bprashanth** [kubernetes/kubernetes#32323]
During node upgrade it looks like sometimes the node can leave the cloudprovider, leading the nodecontroller to think it should delete pods on the node. This is obviously disruptive, but if the pod is not managed by an RC, it can just disappear, which is worse. 

Create a job
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: echo-success
spec:
  template:
    metadata:
      name: echo-success
    spec:
      containers:
      - name: c
        image: busybox
        command: ["/bin/sh",  "-c", "echo success"]
      restartPolicy: Never
  parallelism: 3
  completions: 3
```

Wait for it to succeed and check job:
```
$ k get job 
NAME           DESIRED   SUCCESSFUL   AGE
echo-success   3         3            31m

$ k get po --show-all 
NAME                        READY     STATUS      RESTARTS   AGE       IP            NODE
echo-success-70guk   0/1       Completed   0          13m
echo-success-83k2b   0/1       Completed   0          13m
echo-success-csjo9   0/1       Completed   0          13m
```

Check nodes
```
NAME                           STATUS                     AGE
kubernetes-master              Ready,SchedulingDisabled   30s
kubernetes-minion-group-a9np   Ready                      34s
kubernetes-minion-group-bj5r   Ready                      35s
kubernetes-minion-group-c26b   Ready                      34s
```

Upgrade nodes
```
$ cluster/gce/upgrade.sh -N v1.4.0-beta.0
```

Check nodes
```
NAME                           STATUS                     AGE
kubernetes-master              Ready,SchedulingDisabled   33m
kubernetes-minion-group-a9np   Ready                      11m
kubernetes-minion-group-bj5r   Ready                      9m
kubernetes-minion-group-c26b   Ready                      7m
```

Check jobs
```
$ k get job
NAME           DESIRED   SUCCESSFUL   AGE
echo-success   3         3            29m

$ k get po --show-all -o wide
NAME                 READY     STATUS      RESTARTS   AGE       IP           NODE
echo-success-83k2b   0/1       Completed   0          29m       10.244.1.4   kubernetes-minion-group-a9np
```

This is confusing because the job says 3 but I can only see 1 when I know I saw 3 before upgrade. 
What's worse, I can't see its logs
```
$ k logs echo-success-83k2b
Error response from daemon: No such container: 35935b7065a03f7ac60edf203c1cdc19aaec6696ddbb1036aebf74509457a30d
```

Node controller has
```
I0908 22:08:28.687091       5 deletion_utils.go:273] Recording Deleting Node kubernetes-minion-group-bj5r because it's not present according to cloud provider event message for node kubernetes-minion-group-bj5r
I0908 22:08:33.689852       5 deletion_utils.go:273] Recording Removing Node kubernetes-minion-group-bj5r from NodeController event message for node kubernetes-minion-group-bj5r
I0908 22:08:33.744494       5 deletion_utils.go:273] Recording Deleting all Pods from Node kubernetes-minion-group-bj5r. event message for node kubernetes-minion-group-bj5r
W0908 22:08:55.886551       5 deletion_utils.go:168] Unable to find Node: kubernetes-minion-group-bj5r, deleting all assigned Pods.
```

@kubernetes/goog-cluster @kubernetes/goog-gke fyi for slo around node upgrade
I don't know if we want to do anything about this, marking p0 1.4 for triage.

**Comment by bprashanth** [comment]
https://github.com/kubernetes/kubernetes/issues/32324 for the logs issue

**Comment by bprashanth** [comment]
hmm, i meant to cc @kubernetes/goog-control-plane instead of cluster

**Comment by bprashanth** [comment]
To clarify, I sort of expect that containers will get *restarted* across node upgrade, because whatever. I don't expect pods to get completed deleted.

**Comment by davidopp** [comment]
@gmarek @yujuhong

**Comment by davidopp** [comment]
@erictune in case this is Job-specific

**Comment by davidopp** [comment]
Actually I could have sworn we had a similar issue with 1.2->1.3 but I can't find the issue/bug right now. I think @erictune will probably remember what I'm talking about and can say whether this is a dupe.

**Comment by davidopp** [comment]
#27968 is what I was thinking of.

**Comment by bprashanth** [comment]
Job controller does its thing. No new jobs are created even though completed job disappear. Node controller deletes the pods. It does so because the node disappears from the cloud provider, that's the part that needs investigation. This line:
```
 Recording Deleting Node kubernetes-minion-group-bj5r because it's not present according to cloud provider event message for node kubernetes-minion-group-bj5r
```

* Did the node actually disappear (because as you can see the nodenames are the same before and after upgrade)
* Can we make the nodecontroller smarter so it knows we're upgrading?

**Comment by gmarek** [comment]
This is expected behavior. IIRC @roberthbailey added this logic around v1.0. I'll see what I can do about it.

**Comment by gmarek** [comment]
My guess it takes some time for the cloud provider to recreate the VM, so if NC will wake up in the period when there's no such node, it'll evict everything from those machines.

**Comment by davidopp** [comment]
> My guess it takes some time for the cloud provider to recreate the VM, so if NC will wake up in the period when there's no such node, it'll evict everything from those machines.

Yes, that seems likely. Can we just set a delay between seeing a node has disappeared from the cloud provider, and deleteing it? The node should already be in NotReady state if it has disapeared from the cloud provider, so we should stop scheduling onto it and should have evicted all the pods from it.

**Comment by davidopp** [comment]
Sorry, last thing I wrote made no sense, deleted it.

**Comment by gmarek** [comment]
That's my initial idea and it should work on GCE. I'm not sure how Node upgrades work on different cloud providers though - can we always expect to get VMs with the same names?

**Comment by davidopp** [comment]
I don't think we should change this behavior right now. As you said, node upgrade deletes the VM. Deleting the VM will kill the containers and delete the pods from the node--nothing we do can prevent that. Some users might actually want the pods to reschedule onto a new node in that situation, to minimize disruption. It's hard to have a one-size-fits-all policy here, and I think what we're doing now is not unreasonable.

I think we can address this in 1.5 using forgiveness (#1574). Users can decide for themselves how long of a node outage they are willing to tolerate before a pod should be moved. The simple version would just place a generic "node unreachable" taint, and they would tolerate that taint with some forgiveness period (e.g. longer than the expected node upgrade outage duration if they want their pods to remain bound across node upgrades). A more sophisticated version would have a special "node being upgraded" taint, which would allow the user to have different forgiveness periods for generic unreachable vs. unreachable due to node being upgraded. However, this would require that the system know that the node is being upgraded, which right now isn't the case. It's vaguely reminiscent of the request for an "in maintenance mode" concept (#30272).

As for the specific question about whether it always gets the same name back on GCE, I asked @yujuhong this question back in February for a different reason, and the answer was that kubelet uses whatever name the cloudprovider gives it. I'm actually not sure why GCE gives the same name if you do delete and re-create.

**Comment by bprashanth** [comment]
So basically add something to known issues (https://github.com/kubernetes/kubernetes/issues/32332) saying if you're on a cloudprovider, and you have pods, they will get deleted? i know it's not something new, but it's surprising because my upgrade took < 5m. Or maybe we want some other way to communicate this on a per provider basis, since on prem doesn't suffer. Eitherway my pods just disappearing is a surprise.

**Comment by gmarek** [comment]
I'm also open to any precise suggestions on what should happen in this case - the fact I can't quickly think of any reasonable semantics that will work across cloud providers doesn't mean there's none.

@yujuhong - do you have any ideas from the Node's perspective?

**Comment by bprashanth** [comment]
If I had a website backed by pods and it just stopped working after upgrade I'd be annoyed. 

One option is to defer fixing to a point release. We've always said "use a RC", now we're just saying, "your pods will get deleted across node upgrade, if you didn't listen to us about the RC, you need to actually recreate the pods". 

> As you said, node upgrade deletes the VM. Deleting the VM will kill the containers and delete the pods from the node--nothing we do can prevent that.

Another option is to find a way to prevent "delete the pods from the node". The vm comes back with the same hostname, so can we just do the same 5m logic for pod eviction?

**Comment by gmarek** [comment]
It does in the current GCE updater. Are you sure it's the case for all cloud providers? I don't want to have a special case that will make things worse for everyone other than GCE/GKE.

**Comment by bprashanth** [comment]
It's worse for everyone who changes the hostnames of their nodes (with or without upgrade, this will take 5m and is treated as a netsplit). I think there're a few cases here:     

1. User deletes node
2. Cloudprovider randomly changes nodename
3. Upgrade ends up deleting node    
    a. It comes back with same nodename
    b. it comes back with different nodename

Ux is worse in 3.b, because it'll take 5m to get working pods post upgrade. It is basically treated as a netsplit, because hostname changed. 

(Ux is also worse in 1, but do users actually care about 5m? they deleted the node, if they want to evict asap, they can delete pods too right)

@justinsb how does AWS handle upgrade?

**Comment by gmarek** [comment]
Waiting has general problem of slowing down the reaction for most common cases, i.e. cluster scaling. If we make it 5 minutes then user will need to wait for 5 minutes before we actually recreate his pods that are backed up by RC or something. Personally I'd eat the problematic upgrade frog to keep the cluster scale-down working relatively quickly.

**Comment by davidopp** [comment]
The proper way to do node upgrade, regardless of platform, is to drain the node first. This will ensure the pods are given termination notice. We have a command to do this, `kubectl drain`. If there's anything we should change in the current behavior, it's to call `kubectl drain` from upgrade.sh and wait for the drain to finish before doing the upgrade. Improving the OSS upgrade experience is on my team's roadmap but we're not going to be able to work on it for 1.5 (we'd love for someone in the community to do this...). Another thing that's on the roadmap (for the cluster-lifecycle folks) is in-place upgrades, which is an even better solution because then you don't need to delete the VM at all, and everything just keeps running as-is.

Independent of this issue, we need to make it clear to people that they need to run their pods under a controller. I've talked with the documentation folks and it's on their TODO list; unfortunately we have a lot of documentation that tells people to create pods directly, but really nobody should be doing that. Node upgrade isn't the only scenario where your pods can get deleted.

**Comment by bprashanth** [comment]
Yeah i agree, and no node upgrade isn't the only situation, but node upgrade is a sure edge where this will happen now which is just going to make our version skew issues worse. I'm fine with clarity and best practices solution.

**Comment by bprashanth** [comment]
Unsurprisingly we create the nodes with "delete boot disk", so on upgrade local data on boot disk is lost (logs/images). Good think we didn't push my local storage proposal for 1.4.

**Comment by bprashanth** [comment]
@kubernetes/sig-cluster-lifecycle do the default upgrade scripts for each cloudprovider still delete nodes in order to upgrade them? do we drain them first as mentiond in David's previous comment? @jingxu97 had some concerns about mounted volumes through this process

**Comment by davidopp** [comment]
`upgrade.sh` uses GCE MIG rolling update, which deletes the VM
AFAIK there is no "default upgrade script" for any other cloud provider.

`upgrade.sh` is not integrated with drain yet.

**Comment by bprashanth** [comment]
Thanks for the clarifiction. I'm also incline to say this *isn't* a big
issue for mounted pvs, because deleting the node is presumably going to
shut it down first, and then detach volumes, not detach volumes while pods
are writing to them (definitely verify this, though). Of course the
shutdown could still send a sigkill, resulting in app layer corruption.

**Comment by jingxu97** [comment]
@davidopp, is there some reason for not integrating the drain in upgrade? For example, controller is not sure where to reschedule the pods because all the nodes might be upgrading at the same time?

**Comment by davidopp** [comment]
No, there is no reason not to do it. cluster-lifecycle SIG owns upgrade.sh and if they have the cycles to add proper drain to upgrade.sh (perhaps by shelling out to "kubectl drain" or at least reusing the code from there), that would be great.

@roberthbailey

**Comment by bprashanth** [comment]
Since we use the mig rolling update, do we actually know which node to
drain before the mig kicks in and picks one for update?

**Comment by davidopp** [comment]
Oh right, I forgot about about that.

We've talked about two possible solutions to this:

(1) Create "node drain as a service" a.k.a. "move kubectl drain into the server" -- an API object where you specify which nodes you want to drain, with an associated controller that figures out a good "schedule" for draining them subject to PDB and any other constraints, and which does the drain (evictions) and then when a node is drained calls out to some endpoint to do the maintenance action, in this case presumably calling out to some GCP endpoint that will then execute the node upgrade. Today's upgrade.sh logic would be split between this controller and the GCP logic that executes the node upgrade.

(2) Modify the MIG rolling updater so that it is GKE-aware, e.g. drains the nodes before upgrading and takes PodDisruptionBudget into account. Today's upgrade.sh could continue to operate more or less as-is, at least in the sense of relying on GCP to take care of scheduling and executing the upgrades.

Similar issues of choosing the node, draining, and respecting PDB come up elsewhere too (e.g. preemptible VMs, cluster auto-scaler scale down, and rescheduler).
